### ISSUE TITLE
Assertion `key_part->fieldnr < altered_table->s->fields' failed.

### ISSUE THREAD

**Issue by RoelVdP** [facebook/mysql-5.6#99]
```
Version: '5.6.24-debug'  socket: '/sda/FB-mysql-5.6.24-linux-x86_64-dbg/socket.sock'  port: 18994  MySQL Community Server (GPL)
mysqld: /git/FB-mysql-5.6/storage/innobase/handler/handler0alter.cc:362: virtual enum_alter_inplace_result ha_innobase::check_if_supported_inplace_alter(TABLE*, Alter_inplace_info*): Assertion `key_part->fieldnr < altered_table->s->fields' failed.
05:01:22 UTC - mysqld got signal 6 ;
```

```
+bt
#0  0x00007f08f0c3a771 in __pthread_kill (threadid=<optimized out>, signo=6) at ../nptl/sysdeps/unix/sysv/linux/pthread_kill.c:61
#1  0x0000000000dc8bf2 in my_write_core (sig=6) at /git/FB-mysql-5.6/mysys/stacktrace.c:422
#2  0x0000000000a49e7c in handle_fatal_signal (sig=6) at /git/FB-mysql-5.6/sql/signal_handler.cc:239
#3  <signal handler called>
#4  0x00007f08eebc85d7 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#5  0x00007f08eebc9cc8 in __GI_abort () at abort.c:90
#6  0x00007f08eebc1546 in __assert_fail_base (fmt=0x7f08eed11128 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=assertion@entry=0x1698a38 "key_part->fieldnr < altered_table->s->fields", file=file@entry=0x16989d0 "/git/FB-mysql-5.6/storage/innobase/handler/handler0alter.cc", line=line@entry=362, function=function@entry=0x169b420 <ha_innobase::check_if_supported_inplace_alter(TABLE*, Alter_inplace_info*)::__PRETTY_FUNCTION__> "virtual enum_alter_inplace_result ha_innobase::check_if_supported_inplace_alter(TABLE*, Alter_inplace_info*)") at assert.c:92
#7  0x00007f08eebc15f2 in __GI___assert_fail (assertion=0x1698a38 "key_part->fieldnr < altered_table->s->fields", file=0x16989d0 "/git/FB-mysql-5.6/storage/innobase/handler/handler0alter.cc", line=362, function=0x169b420 <ha_innobase::check_if_supported_inplace_alter(TABLE*, Alter_inplace_info*)::__PRETTY_FUNCTION__> "virtual enum_alter_inplace_result ha_innobase::check_if_supported_inplace_alter(TABLE*, Alter_inplace_info*)") at assert.c:101
#8  0x00000000010bd5be in ha_innobase::check_if_supported_inplace_alter (this=0x7f085a44da10, altered_table=0x7f085a572e00, ha_alter_info=0x7f08f1217660) at /git/FB-mysql-5.6/storage/innobase/handler/handler0alter.cc:361
#9  0x0000000000b67fcf in remove_secondary_keys (thd=0x7f08ec9a4000, create_info=0x7f08f12192c0, table=0x7f085a572e00, alter_info=0x7f08f12191e0) at /git/FB-mysql-5.6/sql/sql_table.cc:8112
#10 0x0000000000b6a288 in mysql_alter_table (thd=0x7f08ec9a4000, new_db=0x7f085a45c690 "test", new_name=0x0, create_info=0x7f08f12192c0, table_list=0x7f085a45c128, alter_info=0x7f08f12191e0, order_num=0, order=0x0, ignore=false) at /git/FB-mysql-5.6/sql/sql_table.cc:9001
#11 0x0000000000cb6874 in Sql_cmd_alter_table::execute (this=0x7f085a45c768, thd=0x7f08ec9a4000) at /git/FB-mysql-5.6/sql/sql_alter.cc:319
#12 0x0000000000afb1b6 in mysql_execute_command (thd=0x7f08ec9a4000, statement_start_time=0x7f08f121a988, post_parse=0x7f08f121b9e0) at /git/FB-mysql-5.6/sql/sql_parse.cc:5585
#13 0x0000000000afed2e in mysql_parse (thd=0x7f08ec9a4000, rawbuf=0x7f085a45c010 "ALTER TABLE t1 CHANGE COLUMN d d CHAR(1)NOT NULL", length=48, parser_state=0x7f08f121b330, last_timer=0x7f08f121b9e0, async_commit=0x7f08f121b9df "") at /git/FB-mysql-5.6/sql/sql_parse.cc:7215
#14 0x0000000000af0bb0 in dispatch_command (command=COM_QUERY, thd=0x7f08ec9a4000, packet=0x7f08647eb001 "ALTER TABLE t1 CHANGE COLUMN d d CHAR(1)NOT NULL", packet_length=48) at /git/FB-mysql-5.6/sql/sql_parse.cc:1595
#15 0x0000000000aef474 in do_command (thd=0x7f08ec9a4000) at /git/FB-mysql-5.6/sql/sql_parse.cc:1071
#16 0x0000000000ab75f8 in do_handle_one_connection (thd_arg=0x7f08ec9a4000) at /git/FB-mysql-5.6/sql/sql_connect.cc:1071
#17 0x0000000000ab6ef7 in handle_one_connection (arg=0x7f08ec9a4000) at /git/FB-mysql-5.6/sql/sql_connect.cc:929
#18 0x00007f08f0c35df5 in start_thread (arg=0x7f08f121c700) at pthread_create.c:308
#19 0x00007f08eec891ad in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:113
```

```
# mysqld options required for replay:  --no-defaults --sql_mode=ONLY_FULL_GROUP_BY
DROP DATABASE test;CREATE DATABASE test;USE test;
CREATE TABLE t1(a INT NOT NULL,b INT UNSIGNED NOT NULL,c CHAR(1),d CHAR (1),e BINARY (1) NOT NULL,f VARCHAR(1) NOT NULL,g BLOB NOT NULL,h BLOB,id INT NOT NULL,KEY(b),KEY(e),KEY (id)) ENGINE=RocksDB;
SET SESSION expand_fast_index_creation=ON;
ALTER TABLE t1 CHANGE COLUMN d d CHAR(1)NOT NULL;
```

Ensure server is started without RocksDB. Use --no-defaults --sql_mode=ONLY_FULL_GROUP_BY

**Comment by laurynas-biveinis** [comment]
Roel, I wonder if the testcase fails for Percona Server?
(the testcase could probably replace engine=rocksdb with engine=innodb)

**Comment by santoshbanda** [comment]
Roel, the failing stack trace shows assertion in innodb code, but your repro case uses rocksdb engine. Could you clarify this?

**Comment by RoelVdP** [comment]
@laurynas-biveinis testcase does not fail on Percona-Server-5.6.25-rel73.2-f9f2b02.Linux.x86_64-debug, irrespective of engine used

@santoshbanda Note that the server is started without RocksDB, hence InnoDB is active, yet the testcase uses a RocksDB CREATE. If you enable RocksDB instead of InnoDB, you get "ERROR 1296 (HY000): Got error 194 'Table must have a PRIMARY KEY' from ROCKSDB" on the CREATE.

**Comment by RoelVdP** [comment]
New finding: testcase also crashes if you use InnoDB on FB branch without RocksDB active.

**Comment by RoelVdP** [comment]
Thus, this seems to be a non-RocksDB bug in FB branch, and the bug may be an InnoDB shortcoming triggered only in FB branch, or a FB branch specific bug.

**Comment by jkedgar** [comment]
I have been able to reproduce the failure and I'm in the process of trying to identify what changed to cause it to fail.  This may take some time.

**Comment by jkedgar** [comment]
@RoelVdP, I think I've found the problem.  I need to run it through all the tests and a code review, but hopefully it will be available soon.

**Comment by RoelVdP** [comment]
@jkedgar Great, thanks

**Comment by jkedgar** [comment]
@RoelVdP, I just committed the fix for this - ef621b24cb137df228b732a18eb11b032bddcea0.

**Comment by RoelVdP** [comment]
Great thanks
